### ISSUE TITLE
Replace also the embed_avg?

### ISSUE THREAD

**Issue by amitaie** [lucidrains/vector-quantize-pytorch#45]
Hey, I'm new in VQ subject so maybe i got it wrong but i have a question about the replacement mechanism in the repo. 

When we replace a vector in the codebook with a vector from the batch, shouldn't we also replace the corresponding index in the "embed_avg" (with the same vector) and kind of reset it? It makes to me no sense to average on the old vector on the next iteration.
The same logic imply on the cluster size, why to keep the old cluster size if we now replaced the vector? shouldn't we reset it to 1 or maybe even better, let the next iteration determinate it's start point cluster size (like in the init k-means). otherwise, it can "kill" a spot in the codebook if we will replace it several times in a raw with no good sample.

If I'm completely wrong and there is an info about it in this repo or papers i would love to know.

Thanks a lot for this repo,
Amitai.

**Comment by lucidrains** [comment]
@amitaie hey Amitai! thank you for raising this issue. i am not an expert with this subject matter either, but i think what you said makes sense and made the change. maybe we can leave the issue open so someone who knows the literature better can chime in!

**Comment by lucidrains** [comment]
@amitaie i don't even remember which paper introduced this technique. was it alongside residual vq and soundstream?

**Comment by amitaie** [comment]
> @amitaie i don't even remember which paper introduced this technique. was it alongside residual vq and soundstream?

It is mention in SoundStream paper:

> Second, as proposed in [34], when a codebook vector has not
been assigned any input frame for several batches, we replace
it with an input frame randomly sampled within the current
batch. More precisely, we track the exponential moving average
of the assignments to each vector (with a decay factor of 0.99)
and replace the vectors of which this statistic falls below 2.

Where 34 is Jukebox's paper.

**Comment by lucidrains** [comment]
@amitaie cool! want to try the recent version and let me know if you see any improvements (or not)?

**Comment by amitaie** [comment]
@lucidrains Yeah, i'm working on that.But i'm working on that through training SoundStream so it will take sometime, hopefully i'll have some results in a week.

In the same subject, i saw that in the Hifi-codec they implement the RVQ differently, with nn.embeding, have any idea why and if it makes difference?

**Comment by lucidrains** [comment]
@amitaie the use of `nn.Embedding` makes no difference, but their proposal to use multi-headed RVQ may

**Comment by amitaie** [comment]
For now i don't see loss improvement. I do see that more vectors in the RVQ is being used, but somehow it doesn't effect the loss. I continue to work on that.

On the same subject, here: 
https://github.com/lucidrains/vector-quantize-pytorch/blob/500db140ceb2939a957e66dcf22849e40ff0daa3/vector_quantize_pytorch/vector_quantize_pytorch.py#L304-L306

we sum all the vectors that we quantized to the same cluster and update the embed_avg accordingly. but in the init with kmeans: 

https://github.com/lucidrains/vector-quantize-pytorch/blob/500db140ceb2939a957e66dcf22849e40ff0daa3/vector_quantize_pytorch/vector_quantize_pytorch.py#L244-L246

we just take the embed and put it in the embed_avg. If i understood correctly the use of embed_avg, I think it should be something like:

` self.embed_avg.data.copy_(embed.clone())`
`self.embed_avg.data[cluser_size > 0] = self.embed_avg.data[cluser_size > 0]  * cluser_size[cluser_size > 0] `

Meaning we multiply the embed by the cluster size (if it's not zero).

Am i right? i'm not sure that i got the way that we update the codebook correctly.
